source/extensions/filters/http/common/aws/utility.cc

```diff
@@ -0,0 +1,54 @@
+#include "extensions/filters/http/common/aws/utility.h"
+
+#include "common/common/fmt.h"
+#include "common/common/utility.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace Common {
+namespace Aws {
+
+std::map<std::string, std::string> Utility::canonicalizeHeaders(const Http::HeaderMap& headers) {
+  std::map<std::string, std::string> out;
+  headers.iterate(
+      [](const Http::HeaderEntry& entry, void* context) -> Http::HeaderMap::Iterate {
+        auto* map = static_cast<std::map<std::string, std::string>*>(context);
+        // Pseudo-headers should not be canonicalized
+        if (entry.key().c_str()[0] == ':') {
+          return Http::HeaderMap::Iterate::Continue;
+        }
+        std::string value(entry.value().getStringView());
+        // Remove leading, trailing, and deduplicate repeated ascii spaces
+        absl::RemoveExtraAsciiWhitespace(&value);
+        const auto iter = map->find(entry.key().c_str());
+        // If the entry already exists, append the new value to the end
+        if (iter != map->end()) {
+          iter->second += fmt::format(",{}", value);
+        } else {
+          map->emplace(entry.key().c_str(), value);
+        }
+        return Http::HeaderMap::Iterate::Continue;
+      },
+      &out);
```

Comment:
I have a feeling this is more generally useful (without the quirks below), but it's fine to leave here, just thinking out loud.